src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -12,237 +12,283 @@ import kotlinx.coroutines.flow.*
+     * This function will throw if the server returns unexpected content or the connection is closed.
+     */
+    private suspend fun handshake(socket: TcpSocket, timeout: Duration) {
+        withTimeout(timeout) {
+            val handshakeFlow = linesFlow(socket)
+                .map { json.decodeFromString(ElectrumResponseDeserializer, it) }
+                .filterIsInstance<Either.Right<Nothing, JsonRPCResponse>>()
+                .map { it.value }
+            // Note that since we synchronously wait for the response, we can safely reuse requestId = 0.
+            val ourVersion = ServerVersion()
+            socket.send(ourVersion.asJsonRPCRequest(id = 0).encodeToByteArray(), flush = true)
+            val theirVersion = parseJsonResponse(ourVersion, handshakeFlow.first())
+            require(theirVersion is ServerVersionResponse) { "invalid server version response $theirVersion" }
+            logger.info { "server version $theirVersion" }
+            socket.send(HeaderSubscription.asJsonRPCRequest(id = 0).encodeToByteArray(), flush = true)
+            val header = parseJsonResponse(HeaderSubscription, handshakeFlow.first())
+            require(header is HeaderSubscriptionResponse) { "invalid header subscription response $header" }
+            _notifications.emit(header)
+            _connectionStatus.value = ElectrumConnectionStatus.Connected(theirVersion, header.blockHeight, header.header)
+            logger.info { "server tip $header" }
         }
+    }
 
-        suspend fun sendRequest(request: ElectrumRequest, requestId: Int) {
-            val bytes = request.asJsonRPCRequest(requestId).encodeToByteArray()
-            try {
-                socket.send(bytes, flush = true)
-            } catch (ex: TcpSocket.IOException) {
-                logger.warning { "cannot send to electrum server" }
-                closeSocket(ex)
+    /**
+     * Start a background coroutine that sends/receive on the TCP socket with the remote server.
+     *
+     * This uses the coroutine scope of the [ElectrumClient], not the scope of the caller, and uses supervision
+     * to ensure that its failure doesn't propagate to the parent, which is likely the application's main scope.
+     *
+     * If this job fails or is cancelled, we clean up resources and set the [connectionStatus] to [ElectrumConnectionStatus.Closed].
+     * The wallet application can then decide to automatically reconnect or switch to a different Electrum server.
+     */
+    private fun listen(socket: TcpSocket): ListenJob {
+        // We use a SupervisorJob to ensure that our CoroutineExceptionHandler is used and exceptions don't propagate
+        // to our parent scope: we simply disconnect and wait for the application to initiate a reconnection.
+        val job = scope.launch(CoroutineName("electrum-client") + SupervisorJob() + CoroutineExceptionHandler { _, ex ->
+            logger.warning(ex) { "electrum connection error: " }
+            socket.close()
```

Comment:
We enter this coroutine exception handler if *one* of the child coroutines threw an exception. Instead of automatically cancelling all other child coroutines and propagating the exception up to our parent, we explicitly close the socket and gracefully terminate, which will send a signal to the other child coroutines to cancel themselves without throwing exceptions.